src/calico_manifests.py

```diff
@@ -375,3 +381,62 @@ def config_hash(self) -> str:
         hash = hashlib.sha256()
         hash.update(json_str.encode())
         return hash.hexdigest()
+
+    def status(self) -> FrozenSet[HashableResource]:
+        """Return all installed objects which have a `.status.conditions` attribute.
+
+        Bonus: Log events for daemonsets with unready pods.
+        """
+        installed = self.installed_resources()
+        for obj in installed:
+            if obj.kind == "DaemonSet":
+                ds: DaemonSet = obj.resource
+                if ds.status.numberReady != ds.status.desiredNumberScheduled:
+                    log_events(collect_events(self.client, obj.resource))
+        return frozenset(_ for _ in installed if _.status_conditions)
+
+    def is_ready(self, obj: HashableResource, cond: AnyCondition) -> Optional[bool]:
+        """Determine if the resource is ready."""
+        if not (ready := super().is_ready(obj, cond)):
+            log_events(collect_events(self.client, obj.resource))
+        return ready
+
+
+def by_localtime(event: Event) -> datetime.datetime:
+    """Return the last timestamp of the event in local time."""
+    dt = event.lastTimestamp or datetime.datetime.now(datetime.timezone.utc)
+    return dt.astimezone()
+
+
+def log_events(events: Iterable[Event]) -> None:
+    """Log the events."""
+    for event in sorted(events, key=by_localtime):
+        log.info(
+            "Event %s/%s %s msg=%s",
+            event.involvedObject.kind,
+            event.involvedObject.name,
+            event.lastTimestamp and event.lastTimestamp.astimezone() or "Date not recorded",
+            event.message,
+        )
+
+
+def collect_events(client: Client, resource: AnyResource) -> List[Event]:
+    """Collect events from the resource."""
+    kind: str = resource.kind or type(resource).__name__
+    meta = resource.metadata
+    object_events = list(
+        client.list(
+            Event,
+            namespace=meta.namespace,
+            fields={
+                "involvedObject.kind": kind,
+                "involvedObject.name": meta.name,
+            },
+        )
+    )
+    if kind in ["Deployment", "DaemonSet"]:
+        involved_pods = client.list(
+            Pod, namespace=meta.namespace, labels={MANIFEST_LABEL: meta.name}
+        )
+        object_events += [event for pod in involved_pods for event in collect_events(client, pod)]
```

Comment:
in these manifests, Deployments and DaemonSets create pods whose labels  are `k8s-app=<name of the resource>`.  Let's look up those pods and see what their events are too!

Reply:
Nice recursive call, I love it!